common/src/main/java/com/discordsrv/common/feature/onlinerole/OnlineRoleModule.java

```diff
@@ -0,0 +1,226 @@
+        return "role";
+    }
+
+    @Override
+    protected List<OnlineRoleConfig> configs() {
+        return Collections.singletonList(discordSRV.config().onlineRole);
+    }
+
+    @Override
+    protected @Nullable ISyncResult doesStateMatch(Boolean one, Boolean two) {
+        if (one == two) {
+            return GenericSyncResults.both(one);
+        }
+        return null;
+    }
+
+    @Override
+    protected Task<Boolean> getDiscord(OnlineRoleConfig config, long userId) {
+        DiscordRole role = discordSRV.discordAPI().getRoleById(config.roleId);
+        if (role == null) {
+            return Task.failed(new SyncFail(OnlineRoleResult.ROLE_DOESNT_EXIST));
+        }
+
+        if (!role.getGuild().getSelfMember().canInteract(role)) {
+            return Task.failed(new SyncFail(OnlineRoleResult.ROLE_CANNOT_INTERACT));
+        }
+
+        return role.getGuild().retrieveMemberById(userId)
+                .mapException(RestErrorResponseException.class, t -> {
+                    if (t.getErrorCode() == ErrorResponse.UNKNOWN_MEMBER.getCode()) {
+                        throw new SyncFail(OnlineRoleResult.NOT_A_GUILD_MEMBER);
+                    }
+                    throw t;
+                })
+                .thenApply(member -> member.hasRole(role) ? true : null);
+    }
+
+    @Override
+    protected Task<Boolean> getGame(OnlineRoleConfig config, UUID playerUUID) {
+        DiscordSRVPlayer player = discordSRV.playerProvider().player(playerUUID);
+
+        if (player == null) {
+            return Task.failed(new SyncFail(OnlineRoleResult.PLAYER_NOT_ONLINE));
+        }
+
+        return Task.completed(!discordSRV.isShutdown());
+    }
+
+    @Override
+    protected Task<ISyncResult> applyDiscord(OnlineRoleConfig config, long userId, @Nullable Boolean newState) {
+        DiscordRole role = discordSRV.discordAPI().getRoleById(config.roleId);
+        if (role == null) {
+            return Task.completed(OnlineRoleResult.ROLE_DOESNT_EXIST);
+        }
+
+        if (!role.getGuild().getSelfMember().canInteract(role)) {
+            return Task.completed(OnlineRoleResult.ROLE_CANNOT_INTERACT);
+        }
+
+        return role.getGuild().retrieveMemberById(userId)
+                .then(member -> Boolean.TRUE.equals(newState)
+                        ? member.addRole(role).thenApply(v -> (ISyncResult) GenericSyncResults.ADD_DISCORD)
+                        : member.removeRole(role).thenApply(v -> GenericSyncResults.REMOVE_DISCORD))
+                .mapException(RestErrorResponseException.class, t -> {
+                    if (t.getErrorCode() == ErrorResponse.UNKNOWN_MEMBER.getCode()) {
+                        throw new SyncFail(OnlineRoleResult.NOT_A_GUILD_MEMBER);
+                    }
+                    if (t.getErrorCode() == ErrorResponse.MISSING_PERMISSIONS.getCode()) {
+                        throw new SyncFail(GenericSyncResults.MEMBER_CANNOT_INTERACT);
+                    }
```

Comment:
Missing Permissions means the bot doesn't have the "Manage Roles" permission, which is different from not being able to interact with the member.

Would be neater to check before trying to do the action via `getSelfMember().hasPermission(Permission.MANAGE_ROLES)`. [DiscordPermissionUtil](https://github.com/DiscordSRV/Ascension/blob/ac9c9528144669863bc6a3a494d34c243575f201/common/src/main/java/com/discordsrv/common/util/DiscordPermissionUtil.java#L32) exists for creating prettier permission error messages (but that might need a separate sync cause implementation)